### PR TITLE
Functional tests - Add new test 'Check invoice'

### PR DESCRIPTION
**Pull request by nesrineabdmouleh**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Add new test 'Check invoice'
| Type?             | refacto
| Category?         | TE
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | no
| How to test?      | no need
| Possible impacts? | no

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27059)

### CODE REVIEW

**Comment by matks** [comment]
Hello, the CI on develop branch was failing today because of https://github.com/PrestaShop/PrestaShop/issues/27062

We fixed the issues on develop branch but you need to rebase to make CI pass successfully on your Pull Request

**Review by matthieu-rolland — COMMENTED** [review]
LGTM, (it's still wip?), but rebase needed 👍

File: tests/UI/pages/commonPage.js [hunk 1]
```diff
@@ -465,6 +465,16 @@ class CommonPage {
...
     return download.path();
   }
+
+  /**
+   * Calculate percentage
+   * @param num {number} Number to do the percentage
+   * @param per {number} Percentage value
+   * @returns {Promise<number>}
+   */
+  async percentage(num, per) {
```
Comment (PierreRambaud) [inline tests/UI/pages/commonPage.js]:
```suggestion
  percentage(num, per) {
```

Is async really needed?

File: tests/UI/pages/BO/orders/view.js [hunk 2]
```diff
@@ -279,6 +291,15 @@ class Order extends BOBasePage {
     return this.elementVisible(page, this.returnProductsButton, 2000);
   }
 
+  /**
+   * Click on view invoice button to download the invoice
+   * @param page {Page} Browser tab
+   * @returns {Promise<void>}
+   */
+  async viewInvoice(page) {
```
Comment (boubkerbribri) [inline tests/UI/pages/BO/orders/view.js]: There is already a function that downloads documents, how this one is different ?

File: tests/UI/pages/commonPage.js (hunk 1 above)
Comment (boubkerbribri) [reply tests/UI/pages/commonPage.js]: I don't really know if its needed, But Even yes, I don'tthink commonPage is the right place (it has nothing to do with pages)

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js [hunk 3]
```diff
@@ -0,0 +1,1229 @@
...
+      await expect(textResult).to.be.equal('Update successful');
+    });
+  });
+
+  // Pre-condition - Create 4 products
+  [virtualProduct,
+    customizedProduct,
+    productWithSpecificPrice,
+    productWithEcoTax,
+  ].forEach((product, index) => {
```
Comment (boubkerbribri) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js]:
```suggestion
  [
    virtualProduct,
    customizedProduct,
    productWithSpecificPrice,
    productWithEcoTax,
  ].forEach((product, index) => {
```

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js [hunk 4]
```diff
@@ -0,0 +1,1229 @@
...
+  // Check invoice
+  [
+    {
+      args: {
+        orderRow: 2,
+        deliveryAddress: Address.second,
+        billingAddress: Address.third,
+        product: virtualProduct,
+        productQuantity: 13,
+        tax: `${virtualProduct.tax} %`,
```
Comment (boubkerbribri) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js]: Could you create this combinations before the test and explain what each one mean please ?

File: tests/UI/pages/commonPage.js (hunk 1 above)
Comment (nesrineabdmouleh) [reply tests/UI/pages/commonPage.js]: Yes async is not needed, I will delete the async.
@boubkerbribri , what's the right place ? wdyt ?

File: tests/UI/pages/BO/orders/view.js (hunk 2 above)
Comment (nesrineabdmouleh) [reply tests/UI/pages/BO/orders/view.js]: The function downloadDocument used on documents tab, but this one is used on the heder of the page( Click on view invoice button).

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js (hunk 4 above)
Comment (nesrineabdmouleh) [reply tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js]: These combinations are not only for for creation, on each one we check an invoice with different data.
I will add some comments.

Comment (boubkerbribri) [reply tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js]: We should know what we test here to be able to reproduce it manually.

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js [hunk 5]
```diff
@@ -0,0 +1,1250 @@
...
+
+            const pageTitle = await viewOrderPage.getPageTitle(page);
+            await expect(pageTitle).to.contains(viewOrderPage.pageTitle);
+          });
+        });
+      }
+
+      // 2 - Create invoice
+      describe('Create invoice', async () => {
+        if (index === 0) {
```
Comment (boubkerbribri) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js]: here

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js [hunk 6]
```diff
@@ -0,0 +1,1250 @@
...
+
+          it('should add the product to the cart', async function () {
+            await testContext.addContextItem(this, 'testIdentifier', `addProductToCart${index}`, baseContext);
+
+            const textResult = await viewOrderPage.addProductToCart(page, test.args.productQuantity);
+            await expect(textResult).to.contains(viewOrderPage.successfulAddProductMessage);
+          });
+        }
+
+        if (index === 0) {
```
Comment (boubkerbribri) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js]: here

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js [hunk 7]
```diff
@@ -0,0 +1,1250 @@
...
+
+            const addressToSelect = `${Address.third.id}- ${Address.third.address} ${Address.third.secondAddress} `
+              + `${Address.third.zipCode} ${Address.third.city}`;
+
+            const alertMessage = await viewOrderPage.selectAnotherInvoiceAddress(page, addressToSelect);
+            expect(alertMessage).to.contains(viewOrderPage.successfulUpdateMessage);
+          });
+        }
+
+        if (index === 0 || index === 1) {
```
Comment (boubkerbribri) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js]: here

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js [hunk 8]
```diff
@@ -0,0 +1,1250 @@
...
+
+  // Pre-condition - Create 4 products
+  [
+    virtualProduct,
+    customizedProduct,
+    productWithSpecificPrice,
+    productWithEcoTax,
+  ].forEach((product, index) => {
+    describe(`PRE-TEST: Create product '${product.name}'`, async () => {
+      if (index === 0) {
```
Comment (boubkerbribri) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js:179]: here

File: tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js [hunk 9]
```diff
@@ -0,0 +1,1250 @@
...
+
+          numberOfProducts = await productsPage.resetAndGetNumberOfLines(page);
+          await expect(numberOfProducts).to.be.above(0);
+        });
+      }
+
+      it('should go to add product page', async function () {
+        await testContext.addContextItem(this, 'testIdentifier', `goToAddProductPage${index}`, baseContext);
+
+        if (index === 0) {
```
Comment (boubkerbribri) [inline tests/UI/campaigns/functional/BO/02_orders/01_orders/viewAndEditOrder/11_checkInvoice.js:202]: here

**Review by boubkerbribri — COMMENTED** [review]
There is too much conditions to run (or not) a step, maybe this scenario should be split (it's more difficult to read it like that)

**Review by boubkerbribri — DISMISSED** [review]

**Review by boubkerbribri — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by PierreRambaud** [comment]
Thank you @nesrineabdmouleh

**Pull request merged**